core/src/main/java/org/apache/calcite/rel/rules/FilterJoinRule.java

```diff
@@ -329,7 +330,14 @@ protected void validateJoinFilters(List<RexNode> aboveFilters,
     final Iterator<RexNode> filterIter = joinFilters.iterator();
     while (filterIter.hasNext()) {
       RexNode exp = filterIter.next();
-      if (!predicate.apply(join, joinType, exp)) {
+      int maxIndex = -1;
+      ImmutableList<RexInputRef> inputRefs = RexUtil.getAllInputRef(exp);
+      for (RexInputRef rexInputRef: inputRefs) {
+        if (rexInputRef.getIndex() > maxIndex) {
+          maxIndex = rexInputRef.getIndex();
+        }
+      }
```

Comment:
Let's doc to explain the scenario here, e.g. semi-join, anti-join ...

Reply:
Good suggestion. I update the PR. Thanks.